pysam/libcutils.pyx

```diff
@@ -144,7 +166,7 @@ cdef charptr_to_str_w_len(const char* s, size_t n, encoding=TEXT_ENCODING):
     if PY_MAJOR_VERSION < 3:
         return s[:n]
     else:
-        return s[:n].decode(encoding)
+        return s[:n].decode(encoding, ERROR_HANDLER)
 
 
 cdef bytes charptr_to_bytes(const char* s, encoding=TEXT_ENCODING):
```

Comment:
@jmarshall : Thanks again for tackling this.

Replacing the default with `None` and adding:
```
if encoding is None:
    encoding = TEXT_ENCODING
```
will allow the default to track the global variable.  This isn't so important for encoding, but the same technique will be needed to support per-file error handlers.  After this PR lands, I'd like to try to add the the error handler as an attribute to the base HTSFile class and see how hard it is to get true per-file settings.  Even if users only use the global setting API, this will also allow the error handler to be fixed at the time a file is opened, rather than changing unexpectedly as a side-effect throughout its lifetime.

Reply:
Yes indeed; an earlier version of this added an `errors=None` as an optional argument to `force_str`/etc as well. I guess the futureproofing thing to do would be to change the signatures to e.g.

```python
def charptr_to_str_w_len(const char* s, size_t n, encoding=None, errors=None):
```
and do that fixing up inside these three or four _libcutils.pyx_ functions as you suggest.

Good point about the constructor needing to give the ability to fix these fields so that file is unaffected by later changes to the global setting.

Reply:
I've updated this so all the relevant functions have `encoding=None, errors=None` optional arguments.